### ISSUE TITLE
Introspection of Current Scope

### ISSUE THREAD

**Issue by convexset** [meteor/blaze#55]
There should be functionality for viewing the current scope programatically just as we can for the data context with `Template.currentData()`... For instance:
```
{{#each things}}
   <!-- Template.currentData() equals each of the items in things -->
{{/each}}
```

But with "scope", there isn't a similar `Template.currentScope()`...
```
{{#each x in things}}
   <!-- x is added to the scope; accessible to helpers by passing but... -->
{{/each}}
```

**Comment by convexset** [comment]
Presumably one could obtain the current scope by:
```javascript
Object.keys(Blaze.currentView._scopeBindings)
  .map(function(k) {
    return {
      id: k,
      value: Blaze.currentView._scopeBindings[k].get()
    };
  });
```
... so why not just have that as `Blaze._getCurrentScope()`?

**Comment by mitar** [comment]
True true. Lexical bindings introduced one more layer of confusion. In Blaze Components [I simply merge data context and lexical bindings together](https://github.com/peerlibrary/meteor-blaze-components/blob/master/lib.coffee#L1026), so that in code you get the same scope as you get inside a template (lexical binding shadowing data context).

You could probably use `Blaze._lexicalBindingLookup(view, symbol)` to get a value.

So, my question is: should `currentData` simply return a merged view of lexical bindings with current data also in Blaze. So that it is the same as it is scope inside a template?

**Comment by convexset** [comment]
My sense is context is context, and this "scope"
is "scope". To avoid confusion, they should stay separate lest people wonder who passed something named"@index" into their template.

Blaze._currentScope() or Template._currentScope() both seem fine.

Another peeve of mine is that there is no scope chain, though one could easily be constructed...

**Comment by stubailo** [comment]
> Lexical bindings introduced one more layer of confusion.

I think it might be a good idea to continue the work Slava and I started, but didn't get an opportunity to finish, of making it possible to avoid data context entirely.

My biggest problem with data context was that it was actually _really hard_ to explain to new devs, since everyone knows how variables work in regular programming languages but data context is really different from that and kind of a hidden thing you can't easily see in the code. IMO it would be much better if variables in Blaze followed the set of rules people are already used to, rather than having an implicit data context that can change at any time, for example:

```
{{#with var}}
   {{#myCustomBlock}}
      {{helper}}
   {{/myCustomBlock}}
{{/with}}
```

In this example, you might expect `helper` to be able to access the context from `var`, but you can't if `#myCustomBlock` does anything with the context. Lexical bindings solve this problem, and make it a lot clearer what is defined where.

**Comment by convexset** [comment]
I feel that data context is intuitive and matches function invocation context (i.e.: `this`).

Scope in Blaze, however, behaves really strangely in that items from parent scopes do not get inherited (and shadowed where applicable).

What I mean is...

```
<template name="ParentTemplate">
  {{#let x=5}}
    {{> ChildTemplate}}
  {{/let}}
</template>

<template name="ChildTemplate">
Can't see x here!
</template>
```
But I guess that's fine because `x` is in a parent (and separate) template... and it would be tough to carve out a scope chain because of `Template.dynamic`. Fine.

But then...
```
{{#each wally in people}}
  [S1]
  {{#each p in props}}
    [S2]
    Where's wally?
  {{/each}}
{{/each}}
```
There is complete shadowing of the entire scope in the same template. Granted, there are the intervening "item" layers. But this is something that can be addressed "statically". The scope chain can be inferred completely from looking at the template itself.

S1: `wally` visible
S2: `wally` and `p` visible

The same issue arises with nested `#let` blocks.

**Comment by convexset** [comment]
@dgreensp: *poke* what do you think? (it's been ages since the ET days)

**Comment by stubailo** [comment]
```
{{#each wally in people}}
  [S1]
  {{#each p in props}}
    [S2]
    Where's wally?
  {{/each}}
{{/each}}
```

In this case, wally is accessible.

By the way, in this example:

```
<template name="ParentTemplate">
  {{#let x=5}}
    {{> ChildTemplate}}
  {{/let}}
</template>

<template name="ChildTemplate">
Can't see x here!
</template>
```

Think of a template as a function:

```
function parentTemplate() {
  var x = 5;
  childTemplate();
}

function childTemplate() {
  // of course you don't see x here!
}
```

I feel like it would be way easier if Blaze and JavaScript variables worked the same way, considering templates to be like functions.

By the way, `this` in JavaScript doesn't match how data context in Blaze behaves at all - `this` doesn't get magically passed between function calls.

**Comment by mitar** [comment]
> By the way, this in JavaScript doesn't match how data context in Blaze behaves at all - this doesn't get magically passed between function calls.

Of course it does. It is also known as `window`. ;-) (But it is not passed between method calls.)

**Comment by convexset** [comment]
> In this case, wally is accessible.

@stubailo True... just not in current scope (`Blaze.currentView._scopeBindings`). But then again, that makes sense because one has to go up the chain.

Which brings me back to a variation on the original question. Why not have functionality to lookup everything that is accessible and to enumerate over that.

**Comment by convexset** [comment]
@stubailo: re `this`

Hmmmm...
```
<template name="T">
  {{#each Things}}
    {{> T2}} or {{> T2 this}}
    {{> T3 somethingInThis}} or {{> T3 this.somethingInThis}}
  {{/each}}
</template>
```
is kinda
```
function T() {
  magicLookup(somethingInThis).forEach(function(item) {
    T2.call(this);
    T3.call(this.somethingInThis);
  });
}
```

(Don't like `{{> T2}}`, makes me have to remember how Blaze works. The explicit stuff is more intuitive. I'm not entirely against setting the whole data context.)

**Comment by mitar** [comment]
One reason why I currently like data context is that if data context changes, Blaze templates are not recreated. So you can store local-to-component state in Blaze templates. For example, is a button toggled or not (things you do not want to persist in a database, state which is really just about UI). Passing arguments explicitly would require some changes here, I think, to make it as smooth.

**Comment by convexset** [comment]
There's good and bad in having functionality to look up scope.

It just encourages more implicit stuff. It would be better if helpers (esp. the global ones) were more "pure". Perhaps `Template.instance()` should be discouraged in helpers. But that's an internal engineering decision.

Event handlers, on the other hand, might need to look up the "scope" they are in. Knowing the entire template instance may not help. Since they are associated with some `event.target`, they can look up "scribbles made" on the node itself (like `data-id`-like attributes) and find their bearings that way (i.e.: rooting around in minimongo), but that may not be ideal.

**Comment by convexset** [comment]
I would propose including the following two functions:

```javascript
Blaze.getScopeVariable = function getScopeVariable(name, view = null) {
	if (!view) {
		// call with something like Blaze.getScopeVariable('x')
		view = Blaze.currentView;
		if (!view) {
			// no current view
			return null;
		}
	}

	// get if available
	if (view._scopeBindings[name]) {
		return view._scopeBindings[name].get();
	}

	// move up template tree; no need to look at the enclosing template as
	// there can't be any scope bindings there
	if (!!view.parentView && !view.parentView._templateInstance) {
		return getScopeVariable(name, view.parentView);
	} else {
		return;  // forever undefined
	}
};

Blaze.getScope = function getScope(view = null, scopeContent = {}) {
	if (!view) {
		// start with a call like: Blaze.getScope()
		view = Blaze.currentView;
		if (!view) {
			// no current view
			return null;
		}
	}

	// populate content with scope bindings where available and not shadowed
	Object.keys(view._scopeBindings).forEach(function addToScopeContent(k) {
		if (!scopeContent[k]) {
			scopeContent[k] = view._scopeBindings[k].get();
		}
	});

	// move up template tree; no need to look at the enclosing template as
	// there can't be any scope bindings there
	if (!!view.parentView && !view.parentView._templateInstance) {
		return getScope(view.parentView, scopeContent);
	} else {
		return scopeContent;
	}
};
```

Should I make a PR?

**Comment by convexset** [comment]
I'm also thinking that lookups can be a little friendlier. Current usage...
```javascript
Blaze.currentView.lookup(name)(); // functions
Blaze.currentView.lookup(name); // non-functions
```
... has the following reasonable behavior: template helpers shadows global helpers shadows data context elements.

But in Spacebars (i.e.: markup/the html templates), the following happens: template helpers shadows scope variables shadows global helpers shadows data context elements. Is this intended?

**Comment by mitar** [comment]
So, current things are pretty inconsistent. I am not sure if we should be adding more special methods to address them for now, because we should probably be changing how things work anyway. So my take would be: we leave this as it is for backwards compatibility, for those who already have code, but then introduce a new clean context/scope which is simply just a template instance and this is it. You attach any helper and state variables to it and this it. The rest is left to the template instance to resolve.

You can use `Blaze._lexicalBindingLookup(Blaze.currentView, symbol)` for now, no?

**Comment by aadamsx** [comment]
> to address them for now, because we should probably be changing how things work anyway

Totally agree with @mitar here.  Instead of bandaids, patches, or hacks, let's go back and fundamentally rethink our base approach from the ground up.

**Comment by convexset** [comment]
Ultimately, Blaze requires some intuitive lookup behavior, but the strange lookup order between template helpers, scope variables, global helpers and data context makes things clunky. Granted, it takes horrid practice to put oneself in a position where one one has to depend on that answer for correct behavior. (I only found that out after experimenting in response realizing I did not know and didn't expect it to be documented.)


I'm aware of:
```javascript
Blaze._lexicalBindingLookup(Blaze.currentView, symbol)
```
... but it is horridly ugly. It was one of the reasons for this whole issue. I was trying to learn how to use scope in Blaze and was rather horrified by it.

Anyway, I guess I'll monkey patch on my end.
